### DISCUSSION TITLE
認証機能

### DISCUSSION THREAD

**Discussion by sanoyo** [sanoyo/ultrachat#10] [Ideas]
## やりたいこと
- googleアカウントで認証できる

## ユーザーストーリー
- ユーザーがsign inボタンをクリックする
- sign in ページに遷移する(下記画像のイメージ)

<img width="947" alt="image" src="https://user-images.githubusercontent.com/24674237/227688485-fb989853-9961-43d5-8f96-be5da4c0b5ff.png">

- ダッシュボードページに遷移する

<img width="1383" alt="image" src="https://user-images.githubusercontent.com/24674237/227688866-fbcf555d-f87e-4963-9b59-2f5f6626996d.png">

## 参考プロダクト
https://app.spatial.chat/

**Comment by sanoyo** [comment]
```mermaid
sequenceDiagram
  participant User
  participant Frontend
  participant Backend
  participant AWS Cognito
  participant Google

  User->>+Frontend: Requests login with Google
  Frontend->>+Backend: Forwards login request
  Backend->>+AWS Cognito: Sends login request
  AWS Cognito->>+Google: Sends login request
  Google->>+AWS Cognito: Sends Google sign-in page
  AWS Cognito->>-Google: Receives sign-in page
  Google->>+AWS Cognito: User signs in and grants permission
  AWS Cognito->>+Google: Sends access token request
  Google->>+AWS Cognito: Sends access token
  AWS Cognito->>-Google: Receives access token
  AWS Cognito->>+Backend: Returns authentication result
  Backend->>-Frontend: Receives authentication result
  Frontend->>-User: Displays authentication result

```

**Reply by sanoyo** [reply]
```mermaid
sequenceDiagram
  participant User
  participant AWS Cognito
  participant AWS AppSync
  participant Backend API
  participant Database

  User->>+AWS Cognito: Requests authentication
  AWS Cognito->>+User: Sends login page
  User->>+AWS Cognito: Logs in and receives access token
  AWS Cognito->>+AWS AppSync: Sends GraphQL request with access token
  AWS AppSync->>+Backend API: Sends request with access token
  Backend API->>+Database: Processes request and retrieves data
  Database->>-Backend API: Returns data
  Backend API->>-AWS AppSync: Returns response
  AWS AppSync->>-AWS Cognito: Returns response
  AWS Cognito->>-User: Returns response

```

**Reply by sanoyo** [reply]
## ref
https://dev.classmethod.jp/articles/amazon-cognito-google-social-signin/
https://www.youtube.com/watch?v=wnB6Fq7-tHs

**Reply by sanoyo** [reply]
cognitには、ユーザープールとIDプールが存在する。
ユーザーの認証にはユーザープールを使用する。

**Reply by sanoyo** [reply]
IDトークンの検証はappsyncでできるみたい。
ちなみにrestだとapi gatewayでやれるみたい。
![image](https://user-images.githubusercontent.com/24674237/227703151-e8c55ee2-6720-4e04-ae89-2270db7a613a.png)

**Reply by sanoyo** [reply]
ユーザープールを作成し、appsyncで認証済みのユーザーだけ叩けることは確認済み

これとやっていることは同じ
https://zenn.dev/tabio/articles/appsync-cognito

**Reply by sanoyo** [reply]
上記踏まえるとこんな感じになる想定

```mermaid
sequenceDiagram
  participant Client
  participant AWS Cognito
  participant AWS AppSync
  participant Backend API
  participant Database

  Client->>+AWS Cognito: Request Authentication
  AWS Cognito->>Client: Get token
  Client->>+AWS AppSync: Request graphql with token
  AWS AppSync->>+Backend API: Sends request
  Backend API->>+Database: Processes request and retrieves data
  Database->>-Backend API: Returns data
  Backend API->>-AWS AppSync: Returns response
  AWS AppSync->>-AWS Cognito: Returns response
  AWS Cognito->>-Client: Returns response
```

**Reply by sanoyo** [reply]
@ddddddO 
多分認証はこんな感じになりそう。
まあどっかで話せればとー。

**Reply by ddddddO** [reply]
@sanoyo 
おおお！凄い！！
りょーかい！ちと来週末予定ありで、もしかするとテキストベースで色々聞くかもです🙏

**Reply by sanoyo** [reply]
りょ〜かーい！
俺が好き勝手進めてるだけなんで、良きタイミングで大丈夫ー！

**Reply by ddddddO** [reply]
貼ってくれた動画も観てみた！
https://github.com/sanoyo/ultrachat/discussions/10#discussioncomment-5424407

1. user -> s3: 静的コンテンツをリクエスト・レスポンス
2. フロント -> Amazon Cognite: 認証画面に認証情報入れてもらって（今回はGoogleの認証画面へ遷移）Cogniteへリクエスト。レスポンスとしてtokenが返る（ユーザープール）
3. フロント -> Amazon Cognite: IDプールでAmazonのリソースへアクセス可能となるクレデンシャルが発行される（ユーザープールでの認証が前提）
4. フロント -> AppSync: クレデンシャル使って、ユーザーはAppSyncを利用できるようになる（チャットが使える）

@sanoyo 
ちょっと確認させてくださいな🙇！

- 認証の流れ、↑の理解で合ってそう...？
- 2で認証するために、認証用画面を用意する感じだよね？？多分

**Reply by sanoyo** [reply]
>認証の流れ、↑の理解で合ってそう...？

あってる！

>2で認証するために、認証用画面を用意する感じだよね？？多分

そうね。
メールアドレス、パスワードであればフィールド作ってもろて、googleだったらgoogleのボタン作ってもらう感じだと思う！
また今週もちびちび進めていくわー

**Reply by ddddddO** [reply]
りょーかい！39！！
